File: aplicaciones/frontend/src/utilidades/localization.ts
/**
 * Utilidades de localización para Chile
 */

export interface ChileanDate {
  day: number;
  month: number;
  year: number;
  hour?: number;
  minute?: number;
}

export interface ChileanCurrency {
  amount: number;
  symbol: string;
  formatted: string;
}

export interface ChileanRUT {
  number: string;
  dv: string;
  formatted: string;
  valid: boolean;
}

export class ChileLocalization {
  private timezone = 'America/Santiago';
  private locale = 'es-CL';
  private currency = 'CLP';

  /**
   * Formatear moneda chilena
   */
  formatCurrency(amount: number, showSymbol: boolean = true): string {
    const formatter = new Intl.NumberFormat('es-CL', {
      style: 'currency',
      currency: 'CLP',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    });

    if (showSymbol) {
      return formatter.format(amount);
    } else {
      return new Intl.NumberFormat('es-CL').format(amount);
    }
  }

  /**
   * Formatear número con separadores de miles chilenos
   */
  formatNumber(number: number, decimals: number = 0): string {
    return new Intl.NumberFormat('es-CL', {
      minimumFractionDigits: decimals,
      maximumFractionDigits: decimals,
    }).format(number);
  }

  /**
   * Formatear fecha en formato chileno
   */
  formatDate(date: Date, formatType: 'short' | 'long' | 'datetime' | 'datetime_long' = 'short'): string {
    const options: Intl.DateTimeFormatOptions = {
      timeZone: this.timezone,
    };

    switch (formatType) {
      case 'short':
        options.day = '2-digit';
        options.month = '2-digit';
        options.year = 'numeric';
        break;
      case 'long':
        options.day = 'numeric';
        options.month = 'long';
        options.year = 'numeric';
        break;
      case 'datetime':
        options.day = '2-digit';
        options.month = '2-digit';
        options.year = 'numeric';
        options.hour = '2-digit';
        options.minute = '2-digit';
        break;
      case 'datetime_long':
        options.day = 'numeric';
        options.month = 'long';
        options.year = 'numeric';
        options.hour = '2-digit';
        options.minute = '2-digit';
        break;
    }

    return new Intl.DateTimeFormat('es-CL', options).format(date);
  }

  /**
   * Formatear RUT chileno
   */
  formatRUT(rut: string): string {
    if (!rut) return rut;

    // Limpiar RUT
    const rutClean = rut.replace(/[^0-9kK]/g, '');

    if (rutClean.length < 2) return rut;

    // Separar número y dígito verificador
    const number = rutClean.slice(0, -1);
    const dv = rutClean.slice(-1).toUpperCase();

    // Formatear con puntos
    const numberFormatted = this.formatNumberWithDots(number);

    return `${numberFormatted}-${dv}`;
  }

  /**
   * Formatear número con puntos cada 3 dígitos
   */
  private formatNumberWithDots(number: string): string {
    if (number.length <= 3) return number;

    return number.replace(/\B(?=(\d{3})+(?!\d))/g, '.');
  }

  /**
   * Validar RUT chileno
   */
  validateRUT(rut: string): boolean {
    if (!rut) return false;

    // Limpiar RUT
    const rutClean = rut.replace(/[^0-9kK]/g, '');

    if (rutClean.length < 2) return false;

    // Separar número y dígito verificador
    const number = rutClean.slice(0, -1);
    const dv = rutClean.slice(-1).toUpperCase();

    // Validar que el número sea válido
    if (!/^\d+$/.test(number)) return false;

    // Calcular dígito verificador
    const dvCalculated = this.calculateRUTDV(number);

    return dv === dvCalculated;
  }

  /**
   * Calcular dígito verificador de RUT
   */
  private calculateRUTDV(number: string): string {
    let sum = 0;
    let multiplier = 2;

    for (let i = number.length - 1; i >= 0; i--) {
      sum += parseInt(number[i]) * multiplier;
      multiplier++;
      if (multiplier > 7) {
        multiplier = 2;
      }
    }

    const remainder = sum % 11;
    const dv = 11 - remainder;

    if (dv === 11) return '0';
    if (dv === 10) return 'K';
    return dv.toString();
  }

  /**
   * Obtener hora actual de Chile
   */
  getCurrentChileanTime(): Date {
    return new Date(new Date().toLocaleString('en-US', { timeZone: this.timezone }));
  }

  /**
   * Parsear fecha en formato chileno
   */
  parseChileanDate(dateString: string, formatType: 'short' | 'long' | 'datetime' = 'short'): Date | null {
    try {
      let date: Date;

      switch (formatType) {
        case 'short':
          // DD/MM/YYYY
          const [day, month, year] = dateString.split('/');
          date = new Date(parseInt(year), parseInt(month) - 1, parseInt(day));
          break;
        case 'long':
          // DD de MMMM de YYYY
          const longMatch = dateString.match(/(\d+)\s+de\s+(\w+)\s+de\s+(\d+)/);
          if (!longMatch) return null;
          const [, dayStr, monthStr, yearStr] = longMatch;
          const monthIndex = this.getMonthNames().indexOf(monthStr);
          if (monthIndex === -1) return null;
          date = new Date(parseInt(yearStr), monthIndex, parseInt(dayStr));
          break;
        case 'datetime':
          // DD/MM/YYYY HH:MM
          const [datePart, timePart] = dateString.split(' ');
          const [day2, month2, year2] = datePart.split('/');
          const [hour, minute] = timePart.split(':');
          date = new Date(parseInt(year2), parseInt(month2) - 1, parseInt(day2), parseInt(hour), parseInt(minute));
          break;
        default:
          return null;
      }

      return date;
    } catch (error) {
      return null;
    }
  }

  /**
   * Obtener nombres de meses en español
   */
  getMonthNames(): string[] {
    return [
      'Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio',
      'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'
    ];
  }

  /**
   * Obtener nombres de días de la semana en español
   */
  getWeekdayNames(): string[] {
    return ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo'];
  }

  /**
   * Formatear número de teléfono chileno
   */
  formatPhone(phone: string): string {
    if (!phone) return phone;

    // Limpiar número
    const phoneClean = phone.replace(/[^0-9]/g, '');

    if (phoneClean.length === 9) {
      // Móvil: 9 1234 5678
      return `${phoneClean[0]} ${phoneClean.slice(1, 5)} ${phoneClean.slice(5)}`;
    } else if (phoneClean.length === 8) {
      // Fijo: 2 1234 5678
      return `${phoneClean[0]} ${phoneClean.slice(1, 5)} ${phoneClean.slice(5)}`;
    }

    return phone;
  }

  /**
   * Obtener sectores económicos chilenos
   */
  getChileanSectors(): Record<string, string> {
    return {
      agricultura: 'Agricultura, Silvicultura y Pesca',
      mineria: 'Minería',
      manufactura: 'Manufactura',
      construccion: 'Construcción',
      comercio: 'Comercio',
      transporte: 'Transporte y Almacenamiento',
      alojamiento: 'Alojamiento y Servicios de Comida',
      informacion: 'Información y Comunicaciones',
      financiero: 'Actividades Financieras y de Seguros',
      inmobiliario: 'Actividades Inmobiliarias',
      profesional: 'Actividades Profesionales, Científicas y Técnicas',
      administrativo: 'Actividades de Servicios Administrativos',
      publico: 'Administración Pública',
      educacion: 'Educación',
      salud: 'Actividades de Atención de la Salud',
      arte: 'Artes, Entretenimiento y Recreación',
      otro: 'Otras Actividades de Servicios',
      hogares: 'Actividades de los Hogares',
      organizaciones: 'Organizaciones y Órganos Extraterritoriales'
    };
  }

  /**
   * Obtener tamaños de empresa chilenos
   */
  getCompanySizes(): Record<string, string> {
    return {
      micro: 'Microempresa (1-9 empleados)',
      pequena: 'Pequeña empresa (10-49 empleados)',
      mediana: 'Mediana empresa (50-199 empleados)',
      grande: 'Gran empresa (200+ empleados)'
    };
  }

  /**
   * Obtener regiones de Chile
   */
  getChileanRegions(): Record<string, string> {
    return {
      arica: 'Región de Arica y Parinacota',
      tarapaca: 'Región de Tarapacá',
      antofagasta: 'Región de Antofagasta',
      atacama: 'Región de Atacama',
      coquimbo: 'Región de Coquimbo',
      valparaiso: 'Región de Valparaíso',
      metropolitana: 'Región Metropolitana',
      ohiggins: 'Región del Libertador General Bernardo O\'Higgins',
      maule: 'Región del Maule',
      nuble: 'Región de Ñuble',
      biobio: 'Región del Biobío',
      araucania: 'Región de La Araucanía',
      rios: 'Región de Los Ríos',
      lagos: 'Región de Los Lagos',
      aysen: 'Región Aysén del General Carlos Ibáñez del Campo',
      magallanes: 'Región de Magallanes y de la Antártica Chilena'
    };
  }

  /**
   * Obtener zona horaria de Chile
   */
  getChileanTimezone(): string {
    return this.timezone;
  }

  /**
   * Obtener locale de Chile
   */
  getChileanLocale(): string {
    return this.locale;
  }

  /**
   * Obtener moneda de Chile
   */
  getChileanCurrency(): string {
    return this.currency;
  }
}

// Instancia global del servicio de localización
export const chileLocalization = new ChileLocalization();

// Funciones de conveniencia
export const formatCLP = (amount: number, showSymbol: boolean = true): string => {
  return chileLocalization.formatCurrency(amount, showSymbol);
};

export const formatChileanDate = (date: Date, formatType: 'short' | 'long' | 'datetime' | 'datetime_long' = 'short'): string => {
  return chileLocalization.formatDate(date, formatType);
};

export const formatRUT = (rut: string): string => {
  return chileLocalization.formatRUT(rut);
};

export const validateRUT = (rut: string): boolean => {
  return chileLocalization.validateRUT(rut);
};

export const getChileanTime = (): Date => {
  return chileLocalization.getCurrentChileanTime();
};

export const formatPhone = (phone: string): string => {
  return chileLocalization.formatPhone(phone);
};

export const getChileanSectors = (): Record<string, string> => {
  return chileLocalization.getChileanSectors();
};

export const getCompanySizes = (): Record<string, string> => {
  return chileLocalization.getCompanySizes();
};

export const getChileanRegions = (): Record<string, string> => {
  return chileLocalization.getChileanRegions();
};

File: aplicaciones/backend/app/core/localization.py
"""
Sistema de localización para Chile
"""
from typing import Dict, Any, Optional
from datetime import datetime, timezone
import locale
import pytz
from decimal import Decimal
import re

class ChileLocalization:
    """Servicio de localización para Chile"""
    
    def __init__(self):
        self.timezone = pytz.timezone('America/Santiago')
        self.locale = 'es_CL.UTF-8'
        self.currency = 'CLP'
        
        # Configurar locale
        try:
            locale.setlocale(locale.LC_ALL, self.locale)
        except locale.Error:
            # Fallback a locale genérico
            try:
                locale.setlocale(locale.LC_ALL, 'es_ES.UTF-8')
            except locale.Error:
                locale.setlocale(locale.LC_ALL, 'C')
    
    def format_currency(self, amount: float, show_symbol: bool = True) -> str:
        """Formatear moneda chilena"""
        if show_symbol:
            return f"${amount:,.0f} CLP"
        else:
            return f"{amount:,.0f}"
    
    def format_number(self, number: float, decimals: int = 0) -> str:
        """Formatear número con separadores de miles chilenos"""
        if decimals > 0:
            return f"{number:,.{decimals}f}".replace(',', 'X').replace('.', ',').replace('X', '.')
        else:
            return f"{number:,.0f}".replace(',', '.')
    
    def format_date(self, date: datetime, format_type: str = "short") -> str:
        """Formatear fecha en formato chileno"""
        if date.tzinfo is None:
            date = date.replace(tzinfo=timezone.utc)
        
        # Convertir a zona horaria de Chile
        chile_date = date.astimezone(self.timezone)
        
        if format_type == "short":
            return chile_date.strftime("%d/%m/%Y")
        elif format_type == "long":
            return chile_date.strftime("%d de %B de %Y")
        elif format_type == "datetime":
            return chile_date.strftime("%d/%m/%Y %H:%M")
        elif format_type == "datetime_long":
            return chile_date.strftime("%d de %B de %Y a las %H:%M")
        else:
            return chile_date.strftime(format_type)
    
    def format_rut(self, rut: str) -> str:
        """Formatear RUT chileno"""
        if not rut:
            return rut
        
        # Limpiar RUT
        rut_clean = re.sub(r'[^0-9kK]', '', rut)
        
        if len(rut_clean) < 2:
            return rut
        
        # Separar número y dígito verificador
        numero = rut_clean[:-1]
        dv = rut_clean[-1].upper()
        
        # Formatear con puntos
        numero_formateado = self._format_number_with_dots(numero)
        
        return f"{numero_formateado}-{dv}"
    
    def _format_number_with_dots(self, number: str) -> str:
        """Formatear número con puntos cada 3 dígitos"""
        if len(number) <= 3:
            return number
        
        result = ""
        for i, digit in enumerate(reversed(number)):
            if i > 0 and i % 3 == 0:
                result = "." + result
            result = digit + result
        
        return result
    
    def validate_rut(self, rut: str) -> bool:
        """Validar RUT chileno"""
        if not rut:
            return False
        
        # Limpiar RUT
        rut_clean = re.sub(r'[^0-9kK]', '', rut)
        
        if len(rut_clean) < 2:
            return False
        
        # Separar número y dígito verificador
        numero = rut_clean[:-1]
        dv = rut_clean[-1].upper()
        
        # Validar que el número sea válido
        try:
            int(numero)
        except ValueError:
            return False
        
        # Calcular dígito verificador
        dv_calculado = self._calculate_rut_dv(numero)
        
        return dv == dv_calculado
    
    def _calculate_rut_dv(self, numero: str) -> str:
        """Calcular dígito verificador de RUT"""
        suma = 0
        multiplicador = 2
        
        for digit in reversed(numero):
            suma += int(digit) * multiplicador
            multiplicador += 1
            if multiplicador > 7:
                multiplicador = 2
        
        resto = suma % 11
        dv = 11 - resto
        
        if dv == 11:
            return "0"
        elif dv == 10:
            return "K"
        else:
            return str(dv)
    
    def get_chilean_timezone(self) -> pytz.timezone:
        """Obtener zona horaria de Chile"""
        return self.timezone
    
    def get_current_chilean_time(self) -> datetime:
        """Obtener hora actual de Chile"""
        return datetime.now(self.timezone)
    
    def parse_chilean_date(self, date_string: str, format_type: str = "short") -> Optional[datetime]:
        """Parsear fecha en formato chileno"""
        try:
            if format_type == "short":
                # DD/MM/YYYY
                return datetime.strptime(date_string, "%d/%m/%Y").replace(tzinfo=self.timezone)
            elif format_type == "long":
                # DD de MMMM de YYYY
                return datetime.strptime(date_string, "%d de %B de %Y").replace(tzinfo=self.timezone)
            elif format_type == "datetime":
                # DD/MM/YYYY HH:MM
                return datetime.strptime(date_string, "%d/%m/%Y %H:%M").replace(tzinfo=self.timezone)
            else:
                return datetime.strptime(date_string, format_type).replace(tzinfo=self.timezone)
        except ValueError:
            return None
    
    def get_month_names(self) -> Dict[int, str]:
        """Obtener nombres de meses en español"""
        return {
            1: "Enero", 2: "Febrero", 3: "Marzo", 4: "Abril",
            5: "Mayo", 6: "Junio", 7: "Julio", 8: "Agosto",
            9: "Septiembre", 10: "Octubre", 11: "Noviembre", 12: "Diciembre"
        }
    
    def get_weekday_names(self) -> Dict[int, str]:
        """Obtener nombres de días de la semana en español"""
        return {
            0: "Lunes", 1: "Martes", 2: "Miércoles", 3: "Jueves",
            4: "Viernes", 5: "Sábado", 6: "Domingo"
        }
    
    def format_phone(self, phone: str) -> str:
        """Formatear número de teléfono chileno"""
        if not phone:
            return phone
        
        # Limpiar número
        phone_clean = re.sub(r'[^0-9]', '', phone)
        
        if len(phone_clean) == 9:
            # Móvil: 9 1234 5678
            return f"{phone_clean[0]} {phone_clean[1:5]} {phone_clean[5:]}"
        elif len(phone_clean) == 8:
            # Fijo: 2 1234 5678
            return f"{phone_clean[0]} {phone_clean[1:5]} {phone_clean[5:]}"
        else:
            return phone
    
    def get_chilean_sectors(self) -> Dict[str, str]:
        """Obtener sectores económicos chilenos"""
        return {
            "agricultura": "Agricultura, Silvicultura y Pesca",
            "mineria": "Minería",
            "manufactura": "Manufactura",
            "construccion": "Construcción",
            "comercio": "Comercio",
            "transporte": "Transporte y Almacenamiento",
            "alojamiento": "Alojamiento y Servicios de Comida",
            "informacion": "Información y Comunicaciones",
            "financiero": "Actividades Financieras y de Seguros",
            "inmobiliario": "Actividades Inmobiliarias",
            "profesional": "Actividades Profesionales, Científicas y Técnicas",
            "administrativo": "Actividades de Servicios Administrativos",
            "publico": "Administración Pública",
            "educacion": "Educación",
            "salud": "Actividades de Atención de la Salud",
            "arte": "Artes, Entretenimiento y Recreación",
            "otro": "Otras Actividades de Servicios",
            "hogares": "Actividades de los Hogares",
            "organizaciones": "Organizaciones y Órganos Extraterritoriales"
        }
    
    def get_company_sizes(self) -> Dict[str, str]:
        """Obtener tamaños de empresa chilenos"""
        return {
            "micro": "Microempresa (1-9 empleados)",
            "pequena": "Pequeña empresa (10-49 empleados)",
            "mediana": "Mediana empresa (50-199 empleados)",
            "grande": "Gran empresa (200+ empleados)"
        }
    
    def get_chilean_regions(self) -> Dict[str, str]:
        """Obtener regiones de Chile"""
        return {
            "arica": "Región de Arica y Parinacota",
            "tarapaca": "Región de Tarapacá",
            "antofagasta": "Región de Antofagasta",
            "atacama": "Región de Atacama",
            "coquimbo": "Región de Coquimbo",
            "valparaiso": "Región de Valparaíso",
            "metropolitana": "Región Metropolitana",
            "ohiggins": "Región del Libertador General Bernardo O'Higgins",
            "maule": "Región del Maule",
            "nuble": "Región de Ñuble",
            "biobio": "Región del Biobío",
            "araucania": "Región de La Araucanía",
            "rios": "Región de Los Ríos",
            "lagos": "Región de Los Lagos",
            "aysen": "Región Aysén del General Carlos Ibáñez del Campo",
            "magallanes": "Región de Magallanes y de la Antártica Chilena"
        }

# Instancia global del servicio de localización
chile_localization = ChileLocalization()

def get_chile_localization() -> ChileLocalization:
    """Obtener instancia del servicio de localización"""
    return chile_localization

# Funciones de conveniencia
def format_clp(amount: float, show_symbol: bool = True) -> str:
    """Formatear moneda chilena"""
    return chile_localization.format_currency(amount, show_symbol)

def format_chilean_date(date: datetime, format_type: str = "short") -> str:
    """Formatear fecha en formato chileno"""
    return chile_localization.format_date(date, format_type)

def format_rut(rut: str) -> str:
    """Formatear RUT chileno"""
    return chile_localization.format_rut(rut)

def validate_rut(rut: str) -> bool:
    """Validar RUT chileno"""
    return chile_localization.validate_rut(rut)

def get_chilean_time() -> datetime:
    """Obtener hora actual de Chile"""
    return chile_localization.get_current_chilean_time()

